compiler/rustc_trait_selection/src/traits/select/candidate_assembly.rs

```diff
@@ -353,7 +355,7 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             all_bounds.filter(|p| p.def_id() == stack.obligation.predicate.def_id());
 
         // Keep only those bounds which may apply, and propagate overflow if it occurs.
-        let mut param_candidates = vec![];
+        let mut param_candidates: SmallVec<[_; 2]> = SmallVec::new();
```

Comment:
why are we even using a separate vec here instead of using
```rust
        for bound in matching_bounds {
            let wc = self.evaluate_where_clause(stack, bound)?;
            if wc.may_apply() {
                candidates.vec.push(ParamCandidate(bound));
            }
        }
```